lib/network/internal.nix

```diff
@@ -0,0 +1,143 @@
+        hasSuffix = strings.hasSuffix "::" addr;
+        hasInfix = strings.hasInfix "::" addr;
+        emptyExpected =
+          if hasPrefix || hasSuffix then
+            2
+          else if hasInfix then
+            1
+          else
+            0;
+      in
+      if addr == "::" then
+        genList (_: "0") ipv6Pieces
+      else if
+        let
+          emptyCount = length pieces - length piecesNoEmpty;
+        in
+        emptyCount != emptyExpected
+        || (hasInfix && piecesNoEmptyLen >= ipv6Pieces)
+        || (!hasInfix && piecesNoEmptyLen != ipv6Pieces)
+      then
+        throw "${apiName}: ${addr} is not a valid IPv6 address"
+      else if hasPrefix || hasSuffix then
+        let
+          zeros = genList (_: "0") (ipv6Pieces - piecesNoEmptyLen);
+        in
+        if hasPrefix then zeros ++ piecesNoEmpty else piecesNoEmpty ++ zeros
+      else if hasInfix then
+        concatMap (
+          piece: if piece == "" then genList (_: "0") (ipv6Pieces - piecesNoEmptyLen) else [ piece ]
+        ) pieces
+      else
+        pieces;
+
+  /**
+    Parses an expanded IPv6 address, converting each part from a string to an
+    integer. Returns a list of integers that can be easily processed by other helper
+    functions.
+  */
+  parseExpandedIpv6 =
+    apiName: addr:
+    let
+      u16FromHexStr =
+        hex:
+        let
+          parsed = trivial.fromHexString hex;
+        in
+        if
+          0 <= parsed && parsed <= 65535 # 2^16 - 1
+        then
+          parsed
+        else
+          throw "${apiName}: 0x${hex} is not a valid u16 integer";
+    in
+    map (piece: u16FromHexStr piece) addr;
+in
+let
+  /**
+    Parses an IPv6 address from a string to - a list of integers.
+  */
+  parseIpv6FromString = apiName: addr: parseExpandedIpv6 apiName (expandIpv6 apiName addr);
+in
+{
+  _ipv6 = {
+    /**
+      Converts an internal representation of an IPv6 address (i.e, a list
+      of integers) to a string. The returned string is not a canonical
+      representation as defined in RFC 5952, zeros are not compressed.
+    */
+    toStringFromExpandedIp =
+      pieces: strings.concatMapStringsSep ":" (piece: strings.toLower (trivial.toHexString piece)) pieces;
+
+    /**
+      Extract an address and subnet prefix length from a string. The subnet
+      prefix length is optional and defaults to 64. The resulting address and prefix
+      length are validated and converted to an internal representation that can be
+      used by other functions.
+    */
+    split =
+      apiName: addr:
+      let
+        splitted = strings.splitString "/" addr;
+        splittedLength = length splitted;
+      in
+      if splittedLength == 1 then # [ ip ]
+        {
+          address = parseIpv6FromString apiName addr;
+          prefixLength = 64; # /64 is a standard size IPv6 subnet as defined by the IETF.
+        }
+      else if splittedLength == 2 then # [ ip subnet ]
```

Comment:
I agree, 128 sounds more logical if prefix length is omitted.

> I think we should be stricter with the user input and don't silently just set some prefix but instead fail saying that it's not a cidr address.

I added one `fromString` function and user can also bot default and cidr notation. I don't see a reason to create two separate API functions like `fromString` and `fromCidrString`, because this can be easily observed from the passed string - if it contains a `/`, then we try to parse it as cidr notation.

Reply:
Ok, when I started working on ipv6 types, I must have understood why we need separate functions for default and cidr notation. In progress.